0000-richer-documents.rst

```diff
@@ -0,0 +1,460 @@
+embed bits of expressions, types and more, in their AST form, right into
+the documents, even supporting embedding ASTs from all 3 GHC passes. We could
+for example "categorise" error annotations according to the type of entity
+(expression, type, ...) and the AST phase (parser, renamer, typechecker):::
+
+    data PassWrapper f
+      = WrapperPs (f GhcPs) -- parser AST
+      | WrapperRn (f GhcRn) -- renamer AST
+      | WrapperTc (f GhcTc) -- typechecker AST
+
+    data ErrAnnotation
+      = ErrHsExpr (PassWrapper HsExpr)
+        -- ^ expression annotation, from any of the 3 phases
+      | ...
+
+Or alternatively build up a large sum type of all possible types of AST chunks
+that the current error infrastructure emits (as text).::
+
+    data ErrAnnotation
+      = ErrTyCon TyCon
+      -- ^ type constructor annotation, used in "Couldn't match expected type"
+      --   error documents and possibly others
+      | ...
+
+The exact definition of this type will be best figured out while implementing
+this proposal, informed by all the error documents that GHC builds.
+
+We could then start emitting ``ErrAnnotation`` values in our documents
+instead of rendering expressions, types, and friends directly.
+GHC's error storage and reporting infrastructure could then be updated to
+emit ``SDoc ErrAnnotation`` values, with the nice side effect that GHC API users
+(such as developers of IDE tooling) would now get to deal with error documents
+with that type. Most users will quite likely then want to use ``>>=`` to process
+those annotations and render them using GHC's default rendition or a custom
+one:::
+
+    (>>=) :: SDoc a -> (a -> SDoc b) -> SDoc b
+
+This ``Monad`` instance lets us substitute all the annotations in a document
+by fresh, new subdocuments that can depend on the annotation values. Of
+particular interest is the case where ``b`` is ``()`` or ``Void``, indicating
+the "annotation-free" nature of the resulting document.  We could even imagine
+tweaking the functions that actually print documents to only accept
+annotation-free documents, so as to force users to interpret annotations one
+way or another before getting the documents printed somewhere.
+
+Annotation-free documents
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In fact, GHC uses ``SDoc`` in other contexts than error messages, a major
+one being code generation. In those cases, we will never want to emit
+any annotation, just pure text, and as fast as possible. This would be another
+case where we would want to deal with ``SDoc Void`` values. It is therefore
+desired that any potential implementation of this proposal doesn't drastically
+change the performance of document construction and rendering when no annotation
+is involved.
+
+Pieces of code that don't produce annotations in documents can either
+produce documents that are completely polymorphic in the annotation type,
+or ``SDoc Void`` values. The polymorphic variant has the merit of being
```

Comment:
Consumers should probably work with `SDoc Void`; producers should probably produce `SDoc a`.

Reply:
Yes, that should be stated clearly, thanks!